lib/puppet/util/windows/adsi.rb

```diff
@@ -176,8 +193,16 @@ def exists?(name_or_sid)
         well_known = false
         if (sid = Puppet::Util::Windows::SID.name_to_principal(name_or_sid))
           # Examples of SidType include SidTypeUser, SidTypeGroup
-          return true if sid.account_type == "SidType#{@object_class.capitalize}".to_sym
-  
+          if sid.account_type == "SidType#{@object_class.capitalize}".to_sym
+            # Check if we're getting back a local user when domain-joined
+            case Puppet::Util::Windows::ADSI.domain_role
+            when :MEMBER_WORKSTATION, :MEMBER_SERVER
+              return sid.domain == Puppet::Util::Windows::ADSI.computer_name
```

Comment:
could this check be used also for other values of `Puppet::Util::Windows::ADSI.domain_role`?

Reply:
I don't think so. We can only have both local and AD accounts when the computer is a member of the domain, so `:MEMBER_WORKSTATION` or `:MEMBER_SERVER`.
For the other cases:
- STANDALONE: should only have local accounts (domain == computer_name)
- CONTROLLER: should only have AD accounts (here the domain doesn't match the computer name, it's the chosen AD domain)

By only handling the affected cases we minimize the risk of possibly breaking something that worked before.